packages/flutter/lib/src/cupertino/nav_bar.dart

```diff
@@ -50,6 +63,76 @@ const TextStyle _kLargeTitleTextStyle = TextStyle(
   color: CupertinoColors.black,
 );
 
+// There's a single tag for all instances of navigation bars because they can
+// all transition between each other (per Navigator) via Hero transitions.
+const _HeroTag _heroTag = _HeroTag();
+
+class _HeroTag {
+  const _HeroTag();
+  // This class exists because the Gallery smoke test wants all widgets from
+  // debugDumpApp() to have a toString();
+  @override
+  String toString() => "Cupertino navigation bars' Hero tag";
+}
+
+TextStyle _navBarItemStyle(Color color) {
+  return new TextStyle(
+    fontFamily: '.SF UI Text',
+    fontSize: 17.0,
+    letterSpacing: -0.24,
+    color: color,
+  );
+}
+
+/// Returns `child` wrapped with background and a bottom border if background color
+/// is opaque. Otherwise, also blur with [BackdropFilter].
+Widget _wrapWithBackground({
+  Border border,
+  Color backgroundColor,
+  Widget child,
+  bool annotate = true,
+}) {
+  Widget result = child;
+  if (annotate) {
+    final bool darkBackground = backgroundColor.computeLuminance() < 0.179;
+    final SystemUiOverlayStyle overlayStyle = darkBackground
+        ? SystemUiOverlayStyle.light
+        : SystemUiOverlayStyle.dark;
+    result = new AnnotatedRegion<SystemUiOverlayStyle>(
+      value: overlayStyle,
+      sized: true,
+      child: result,
+    );
+  }
+  final DecoratedBox childWithBackground = new DecoratedBox(
+    decoration: new BoxDecoration(
+      border: border,
+      color: backgroundColor,
+    ),
+    child: result,
+  );
+
+  if (backgroundColor.alpha == 0xFF)
+    return childWithBackground;
+
+  return new ClipRect(
+    child: new BackdropFilter(
+      filter: new ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
+      child: childWithBackground,
+    ),
+  );
```

Comment:
clips and backdrop filters are expensive :-(

cc @liyuqian in case you have any ideas for how else to do this

Reply:
We had this conversation in the diffbase :D
https://github.com/flutter/flutter/pull/19637#pullrequestreview-142598550

I just moved stuff around.

Reply:
Clipping rect should be fairly cheap as long as the rect is axis-aligned. You can also set `clipBehavior = Clip.hardEdge` to further reduce the performance overhead. I wouldn't expect such a clip rect to become a performance bottleneck.

Reply:
@liyuqian: does hardEdge help performance further if the rect is already axis-aligned? Otherwise I'd be cautious to 'reduce fidelity' then add another API for users to boost it back if they put the nav bar in a rotate or some such.

Reply:
I think it's fine to do that. It's what we've done pretty much everywhere else for similar things now.

Reply:
ah ya, you're right. I didn't see the refactor PR. It's like all over the place now...
ok, will do

Reply:
Unfortunately I had to revert this since it causes a number of graphical bugs. For instance, having a hardEdge clip would cause other clips (hardEdge or not) on the same page to no longer appear. 

It also causes https://github.com/flutter/flutter/issues/21095.

I'll create reduced test cases and file bugs for those issues separately.